main/metrics.c

```diff
@@ -0,0 +1,94 @@
+    METRICS_PREFIX "heap_free_bytes{capability=\"MALLOC_CAP_DEFAULT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_EXEC\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_32BIT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_8BIT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_DMA\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID2\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID3\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID4\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID5\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID6\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID7\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_SPIRAM\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_INTERNAL\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_DEFAULT\"} %zu\n",
+    gw_metrics.received_advertisements,
+    esp_timer_get_time(),
+    get_total_free_bytes(MALLOC_CAP_EXEC),
+    get_total_free_bytes(MALLOC_CAP_32BIT),
+    get_total_free_bytes(MALLOC_CAP_8BIT),
+    get_total_free_bytes(MALLOC_CAP_DMA),
+    get_total_free_bytes(MALLOC_CAP_PID2),
+    get_total_free_bytes(MALLOC_CAP_PID3),
+    get_total_free_bytes(MALLOC_CAP_PID4),
+    get_total_free_bytes(MALLOC_CAP_PID5),
+    get_total_free_bytes(MALLOC_CAP_PID6),
+    get_total_free_bytes(MALLOC_CAP_PID7),
+    get_total_free_bytes(MALLOC_CAP_SPIRAM),
+    get_total_free_bytes(MALLOC_CAP_INTERNAL),
+    get_total_free_bytes(MALLOC_CAP_DEFAULT),
+    heap_caps_get_largest_free_block(MALLOC_CAP_EXEC),
+    heap_caps_get_largest_free_block(MALLOC_CAP_32BIT),
+    heap_caps_get_largest_free_block(MALLOC_CAP_8BIT),
+    heap_caps_get_largest_free_block(MALLOC_CAP_DMA),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID2),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID3),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID4),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID5),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID6),
+    heap_caps_get_largest_free_block(MALLOC_CAP_PID7),
+    heap_caps_get_largest_free_block(MALLOC_CAP_SPIRAM),
+    heap_caps_get_largest_free_block(MALLOC_CAP_INTERNAL),
+    heap_caps_get_largest_free_block(MALLOC_CAP_DEFAULT));
+}
+
+char* ruuvi_get_metrics()
+{
+    char *buf = malloc(BUF_LEN * sizeof(char));
+    int size = gen_metrics(buf, BUF_LEN);
+    if (size >= BUF_LEN) {
+        ESP_LOGW (TAG, "Initial buffer size of %d for metrics is insufficient, needed %d! Consider increasing BUF_LEN in metrics.c", BUF_LEN, size);
+        realloc(buf,(size + 1) * sizeof(char));
```

Comment:
I think realloc may move memory to a new location if the existing pointer cannot be expanded to fit new size and free the old pointer. New value of buffer should be stored and used.

Reply:
Good catch, I forgot realloc doesn't (necessarily) just modify the original one. I'll fix this when the decision is done where to allocate this. I think it's good to have a consistent way of doing things, but indeed it would be clearer to do both the allocation and freeing on the same side.

Reply:
there is a better way - calculate the required size for the buffer, then allocate it. I'll do a refactoring.